File: flutter_app/lib/models/activity_log.dart
class ActivityLog {
  final String id;
  final String userId;
  final String action;
  final String entity;
  final String details;
  final String description;
  final String userName;
  final String userRole;
  final String ipAddress;
  final String device;
  final String browser;
  final DateTime timestamp;
  final DateTime createdAt;

  ActivityLog({
    required this.id,
    required this.userId,
    required this.action,
    required this.entity,
    required this.details,
    required this.description,
    required this.userName,
    required this.userRole,
    required this.ipAddress,
    required this.device,
    required this.browser,
    required this.timestamp,
    required this.createdAt,
  });

  factory ActivityLog.fromJson(Map<String, dynamic> json) {
    return ActivityLog(
      id: json['_id'] ?? '',
      userId: json['userId'] ?? '',
      action: json['action'] ?? '',
      entity: json['entity'] ?? '',
      details: json['details'] ?? '',
      description: json['description'] ?? json['details'] ?? '',
      userName: json['userName'] ?? 'Unknown',
      userRole: json['userRole'] ?? 'user',
      ipAddress: json['ipAddress'] ?? '',
      device: json['device'] ?? 'Unknown',
      browser: json['browser'] ?? 'Unknown',
      timestamp: DateTime.tryParse(json['timestamp'] ?? '') ?? DateTime.now(),
      createdAt: DateTime.tryParse(json['createdAt'] ?? json['timestamp'] ?? '') ?? DateTime.now(),
    );
  }
}

File: backend/models/ActivityLog.js
const mongoose = require('mongoose');

const activityLogSchema = new mongoose.Schema({
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Faculty',
    required: true
  },
  userName: {
    type: String,
    required: true
  },
  userRole: {
    type: String,
    required: true
  },
  action: {
    type: String,
    required: true,
    enum: [
      'CREATE', 'READ', 'UPDATE', 'DELETE',
      'LOGIN', 'LOGOUT', 'UPLOAD', 'DOWNLOAD',
      'APPROVE', 'REJECT', 'PUBLISH', 'ROLLBACK'
    ]
  },
  entity: {
    type: String,
    required: true,
    enum: [
      'Faculty', 'Subject', 'Section', 'Room', 'Timetable',
      'Department', 'Announcement', 'University', 'User'
    ]
  },
  entityId: {
    type: String,
    default: null
  },
  description: {
    type: String,
    required: true
  },
  changes: {
    type: mongoose.Schema.Types.Mixed,
    default: {}
  },
  ipAddress: {
    type: String,
    default: ''
  },
  userAgent: {
    type: String,
    default: ''
  },
  device: {
    type: String,
    default: ''
  },
  browser: {
    type: String,
    default: ''
  },
  status: {
    type: String,
    enum: ['success', 'failure'],
    default: 'success'
  },
  errorMessage: {
    type: String,
    default: null
  }
}, {
  timestamps: true
});

// Indexes for faster queries
activityLogSchema.index({ userId: 1, createdAt: -1 });
activityLogSchema.index({ action: 1, entity: 1 });
activityLogSchema.index({ createdAt: -1 });

// Auto-delete logs older than 90 days (optional)
activityLogSchema.index({ createdAt: 1 }, { expireAfterSeconds: 7776000 });

module.exports = mongoose.model('ActivityLog', activityLogSchema);
